python/tank_vendor/shotgun_deploy/descriptor_io/appstore.py

```diff
@@ -0,0 +1,656 @@
+                ["description",
+                 "sg_detailed_release_notes",
+                 "sg_documentation",
+                 constants.TANK_CODE_PAYLOAD_FIELD]
+            )
+            if version is None:
+                raise ShotgunDeployError(
+                    "The App store does not have a version '%s' of Core!" % self._version
+                )
+            
+        else:
+            # engines, apps etc have a 'bundle level entity' in the app store,
+            # e.g. something representing the app or engine.
+            # then a version entity representing a particular version
+            bundle = sg.find_one(
+                bundle_entity,
+                [["sg_system_name", "is", self._name]],
+                ["sg_status_list", "sg_deprecation_message"]
+            )
+
+            if bundle is None:
+                raise ShotgunDeployError(
+                    "The App store does not contain an item named '%s'!" % self._name
+                )
+    
+            # now get the version
+            version = sg.find_one(
+                version_entity,
+                [[link_field, "is", bundle], ["code", "is", self._version]],
+                ["description",
+                 "sg_detailed_release_notes",
+                 "sg_documentation",
+                 constants.TANK_CODE_PAYLOAD_FIELD]
+            )
+            if version is None:
+                raise ShotgunDeployError(
+                    "The App store does not have a "
+                    "version '%s' of item '%s'!" % (self._version, self._name)
+                )
+
+        metadata = {"bundle": bundle, "version": version}
+
+        fp = open(path, "wt")
+        try:
+            pickle.dump(metadata, fp)
+            log.debug("Wrote app store cache file '%s'" % path)
+        finally:
+            fp.close()
+
+        return metadata
+
+    def _get_cache_paths(self):
+        """
+        Get a list of resolved paths, starting with the primary and
+        continuing with alternative locations where it may reside.
+
+        :return: List of path strings
+        """
+        paths = []
+
+        for root in [self._bundle_cache_root] + self._fallback_roots:
+            paths.append(
+                os.path.join(
+                    root,
+                    "app_store",
+                    self.get_system_name(),
+                    self.get_version()
+                )
+            )
+        return paths
+
+    ###############################################################################################
+    # data accessors
+
+    @classmethod
+    def dict_from_uri(cls, uri):
+        """
+        Given a location uri, return a location dict
+
+        :param uri: Location uri string
+        :return: Location dictionary
+        """
+        # sgtk:app_store:tk-core:v12.3.4
+
+        # explode into dictionary
+        location_dict = cls._explode_uri(uri, "app_store", ["name", "version"])
```

Comment:
Shouldn't we have a property on the class that has the same of the type? Then some of this could move to a base class. The get_cache_paths could also move into a base class for versioned descriptors